### PR TITLE
feat(experiments): Internalize experiment rules.

### LINKED ISSUE #4893: Kill the experiments repo
Experiments have served us well, but we are not using the dynamic ability as we initially thought we would. We have never turned off a feature using Able, instead we have always just rolled back. We have never increased the % of users who go through a flow mid-train, instead we have always waited for the next train.

Having experiments in a separate repo introduces complexity in the build step. Let's remove the complexity.

### PR DESCRIPTION
**Pull request by shane-tomlinson**
This allows us to remove Able.

fixes #4893

### CODE REVIEW

**Comment by vladikoff** [comment]
We can also clean up the configuration. It the main configuration file and also there is a `production-experiments.json` file

**Comment by vladikoff** [comment]
@shane-tomlinson let's chat about this one

File: app/tests/spec/views/mixins/experiment-mixin.js [hunk 1]
```diff
@@ -44,6 +44,7 @@ define(function (require, exports, module) {
         trigger: sinon.spy()
       };
       windowMock = new WindowMock();
+      windowMock = new WindowMock();
```
Comment (vladikoff) [inline app/tests/spec/views/mixins/experiment-mixin.js]:
```
windowMock = new WindowMock();
 +      windowMock = new WindowMock();
```

:)

File: server/lib/configuration.js [hunk 2]
```diff
@@ -126,24 +126,6 @@ const conf = module.exports = convict({
       'development'
     ]
   },
-  experiments: {
```
Comment (vladikoff) [inline server/lib/configuration.js]: Is the only part we are missing here is a config via ENV variables, like we were planning for Phase 1?

I am okay with merging this without it, and then make another PR to make stuff like `SMS stuff` configurable. Or we can make SMS 100% and add the ENV switching for some other new feature

**Comment by shane-tomlinson** [comment]
> Is the only part we are missing here is a config via ENV variables, like we were planning for Phase 1?
>
> I am okay with merging this without it, and then make another PR to make stuff like SMS stuff configurable. Or we can make SMS 100% and add the ENV switching for some other new feature

TBH, I kinda hoped we could punt on the ENV variables to try to get this part closed off w/o adding an extra layer of complexity.

File: app/tests/spec/views/mixins/experiment-mixin.js (hunk 1 above)
Comment (shane-tomlinson) [inline app/tests/spec/views/mixins/experiment-mixin.js]: He... I need a bashful face for that one.

File: app/scripts/lib/experiments/choices/README.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+Experiment choice rules decide whether a user is part of an experiment,
+and if so, which group.
+
+## Creating a new rule.
+
+1. Copy [TEMPLATE.js](https://github.com/mozilla/fxa-content-server/tree/master/app/scripts/experiments/rules/TEMPLATE.js) to a new rule file.
+2. Change `ChangeMeExperiment` class name to another name.
+3. Fill in an experiment name for `CHANGE_ME` in the constructor.
+4. Fill in the `choose` function.
+5. Include the new rule file in [index.js](https://github.com/mozilla/fxa-content-server/tree/master/app/scripts/experiments/rules/index.js).
```
Comment (shane-tomlinson) [inline app/scripts/lib/experiments/choices/README.md]: `rules`=>`choices` throughout this file.

File: app/scripts/require_config.js [hunk 4]
```diff
@@ -69,6 +69,7 @@ require.config({
     canvasToBlob: '../bower_components/blueimp-canvas-to-blob/js/canvas-to-blob',
     chai: '../bower_components/chai/chai',
     cocktail: '../bower_components/cocktail/Cocktail',
+    crc: '../bower_components/js-crc/src/crc',
```
Comment (vladikoff) [inline app/scripts/require_config.js]: 👍

File: app/scripts/lib/experiments/choices/send-sms-enabled-for-country.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+define((require, exports, module) => {
+  'use strict';
+
+  const BaseExperiment = require('./base');
+
+  function canEmailSendToRo (email) {
+    return /@softvision\.(com|ro)$/.test(email) ||
+          /@mozilla\.(com|org)$/.test(email);
```
Comment (shane-tomlinson) [inline app/scripts/lib/experiments/choices/send-sms-enabled-for-country.js]: nit: Add a space.

File: app/scripts/lib/experiments/choices/communication-prefs.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+define((require, exports, module) => {
+  'use strict';
+
+  const BaseExperiment = require('./base');
```
Comment (shane-tomlinson) [inline app/scripts/lib/experiments/choices/communication-prefs.js]: nit: `BaseExperiment` should probably be called `BaseChoice` or `BaseChoiceRule`

File: app/scripts/lib/experiments/choices/TEMPLATE.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ * 4. Fill in the `choose` function.
+ * 5. Include the new rule file in index.js.
+ * 6. Access in views via `this.experiments.choose('experiment name from 3')`
+ */
+define((require, exports, module) => {
+  'use strict';
+
+  const BaseExperiment = require('./base');
+
+  module.exports = class ChangeMeExperiment extends BaseExperiment {
```
Comment (shane-tomlinson) [inline app/scripts/lib/experiments/choices/TEMPLATE.js]: nit: `ChangeMeExperiment` should probably be called `ChangeMeExperimentChoice` or `ChangeMeExperimentChoiceRule`

File: app/scripts/lib/experiments/choices/TEMPLATE.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+ * 2. Change `ChangeMeExperiment` class name to another name.
+ * 3. Fill in an experiment name for `CHANGE_ME` in the constructor.
+ * 4. Fill in the `choose` function.
+ * 5. Include the new rule file in index.js.
+ * 6. Access in views via `this.experiments.choose('experiment name from 3')`
+ */
+define((require, exports, module) => {
+  'use strict';
+
+  const BaseExperiment = require('./base');
```
Comment (shane-tomlinson) [inline app/scripts/lib/experiments/choices/TEMPLATE.js]: See other comment about `BaseExperiment`, apply to all of these choices.

File: app/scripts/lib/app-start.js [hunk 9]
```diff
@@ -88,8 +88,8 @@ define(function (require, exports, module) {
       this._interTabChannel = new InterTabChannel();
     },
 
-    initializeAble () {
-      this._able = new Able();
+    initializeExperimentChoices () {
+      this._experimentChoices = new ExperimentChoices();
```
Comment (shane-tomlinson) [inline app/scripts/lib/app-start.js]: Is `experimentChoices` the right name? `experimentChoiceRules` perhaps?

File: app/scripts/lib/experiments/choices/TEMPLATE.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+  module.exports = class ChangeMeExperiment extends BaseExperiment {
+    constructor () {
+      super();
+      this.name = 'CHANGE_ME';
+    }
+
+    choose (subject) {
```
Comment (shane-tomlinson) [inline app/scripts/lib/experiments/choices/TEMPLATE.js]: Add a jsdoc to the top of this method to help devs that write new choice rules.

File: app/scripts/lib/experiments/choices/sentry.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+define((require, exports, module) => {
+  'use strict';
+
+  const BaseExperiment = require('./base');
+
+  module.exports = class SentryExperiment extends BaseExperiment {
```
Comment (vladikoff) [inline app/scripts/lib/experiments/choices/sentry.js]: nit: This experiment is missing a bit of docs. Do you think someone new to the project will be able to understand what this does? I guess they can look at the `sentryEnabled` code too.

File: app/scripts/lib/experiments/choices/index.js [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * Interface to experiment choices.
+ */
+
+define((require, exports, module) => {
+  'use strict';
+
+  const _ = require('underscore');
+
+  const ExperimentChoices = [
```
Comment (vladikoff) [inline app/scripts/lib/experiments/choices/index.js]: Just to confirm, all these are experiments can mix together.

If we wanted to create 2 independent experiments then we would to :

```js
 require('./some-name-experiment-chooser'),
 require('./some-name-experiment-1'),
 require('./some-name-experiment-2')
```

Where `some-name-experiment-chooser` would pick between `1` and `2`?

Correct?

Comment (shane-tomlinson) [inline app/scripts/lib/experiments/choices/index.js]: Can you explain in different words? I'm a bit confused.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 43b067090079a77a01342373f84043022734a239
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

File: app/scripts/lib/experiments/choices/index.js (hunk 12 above)
Comment (shane-tomlinson) [inline app/scripts/lib/experiments/choices/index.js]: Added a test and documentation to show how this works!

File: app/scripts/lib/experiments/choices/sentry.js (hunk 11 above)
Comment (shane-tomlinson) [inline app/scripts/lib/experiments/choices/sentry.js]: Added a tiny blurb to the top of each module

**Comment by shane-tomlinson** [comment]
@vladikoff - I call this ready for review!

File: app/scripts/lib/experiments/grouping-rules/README.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+be used to recursively call other tests.
+
+```js
+choose (subject = {}) {
+  const choice = this.uniformChoice(['recursive-rule-1', 'recursive-rule-2'], subject.uniqueUserId);
+  return subject.experimentGroupingRules.choose(choice, subject);
+}
+```
+
+### Mutually exclusive grouping rules
```
Comment (vladikoff) [inline app/scripts/lib/experiments/grouping-rules/README.md]: 👍 thanks so much!
